src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -10,49 +10,554 @@
+          _archiver_term.append_tx(meta.archiver_term()),
+          _segment_term.append_tx(meta.segment_term()),
+          _delta_offset_end.append_tx(meta.delta_offset_end()),
+          _sname_format.append_tx(static_cast<int64_t>(meta.sname_format))));
+
+        if (
+          ix
+            % static_cast<uint32_t>(::details::FOR_buffer_depth * sampling_rate)
+          == 0) {
+            // At the begining of every row we need to collect
+            // a set of hints to speed up the subsequent random
+            // reads.
+            auto base_offset_hint = _base_offset.get_current_stream_pos();
+            // Invariant: it's guaranteed that if we will get nullopt from
+            // one column we will get nullopt from other columns. The opposite
+            // is also true.
+            if (base_offset_hint.has_value()) {
+                auto tup = std::make_tuple(
+                  *_is_compacted.get_current_stream_pos(),
+                  *_size_bytes.get_current_stream_pos(),
+                  *base_offset_hint,
+                  *_committed_offset.get_current_stream_pos(),
+                  *_base_timestamp.get_current_stream_pos(),
+                  *_max_timestamp.get_current_stream_pos(),
+                  *_delta_offset.get_current_stream_pos(),
+                  *_ntp_revision.get_current_stream_pos(),
+                  *_archiver_term.get_current_stream_pos(),
+                  *_segment_term.get_current_stream_pos(),
+                  *_delta_offset_end.get_current_stream_pos(),
+                  *_sname_format.get_current_stream_pos());
+                _hints.insert(std::make_pair(ix, tup));
+            } else {
+                _hints.insert(std::make_pair(ix, std::nullopt));
+            }
+        }
+    }
+
+    static segment_meta dereference(const iterators_t& it) {
+        segment_meta meta = {};
+        details::value_or<segment_meta_ix::is_compacted>(it, meta.is_compacted);
+        details::value_or<segment_meta_ix::size_bytes>(it, meta.size_bytes);
+        details::value_or<segment_meta_ix::base_offset>(it, meta.base_offset);
+        details::value_or<segment_meta_ix::committed_offset>(
+          it, meta.committed_offset);
+        details::value_or<segment_meta_ix::base_timestamp>(
+          it, meta.base_timestamp);
+        details::value_or<segment_meta_ix::max_timestamp>(
+          it, meta.max_timestamp);
+        details::value_or<segment_meta_ix::delta_offset>(it, meta.delta_offset);
+        details::value_or<segment_meta_ix::ntp_revision>(it, meta.ntp_revision);
+        details::value_or<segment_meta_ix::archiver_term>(
+          it, meta.archiver_term);
+        details::value_or<segment_meta_ix::segment_term>(it, meta.segment_term);
+        details::value_or<segment_meta_ix::delta_offset_end>(
+          it, meta.delta_offset_end);
+        details::value_or<segment_meta_ix::sname_format>(it, meta.sname_format);
```

Comment:
nit: it kind of seems like with the type tricks (ie foo<hint_t, hint_t, ...>) things weren't taken quite far enough if we still have to type all this out.

Reply:
I tried to do this using `Boost.Fusion` but didn't like the result. I will try something different in the next PR.